### PR TITLE
K1p1

### PR DESCRIPTION
**Pull request by graceolivia**
We discussed me doing a pull request so you could take a look at my code so far if you wanted to. 

The main work I've done so far is a chart-printing function. <strike> It can't handle multi-line user input yet </strike>, although there is some code to think about handling it so far. There are ZOM tests for the chart-printing function. It can only handle k1 and p1.

Update: This is updated to handle user input including k1, p1, and \n, which causes a line break. Added a function to check user input and convert it into a 2d array for processing.

### CODE REVIEW

File: Tests/SwiftCliCoreTests/ChartConstructorTest.swift [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+import XCTest
+@testable import SwiftCliCore
+
+
+class TopLevelPrinterTests: XCTestCase {
+    
+    // Top Level Tests
+    func testPrintTopLevelMany() throws {
+        let result = ChartConstructor().make_top_row(width: 4)
+        XCTAssertEqual(result, "┌─┬─┬─┬─┐\n")
```
Comment (hnlee) [inline Tests/SwiftCliCoreTests/ChartConstructorTest.swift:11]: This is a great start on writing tests! If you have time, I would recommend looking at using `Nimble` as a test dependency for more expressive assertions.

File: Tests/SwiftCliCoreTests/InputValidatorTest.swift [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    func testValidatorFor0Input() throws {
+        let result = Validator().validate(pattern: "")
+        XCTAssertEqual(result, false)
+    }
+    
+    func testValidatorForMiscBadInput() throws {
+        let result = Validator().validate(pattern: "3")
+        XCTAssertEqual(result, false)
+    }
+
```
Comment (hnlee) [inline Tests/SwiftCliCoreTests/InputValidatorTest.swift:39]: Very nice!

File: Tests/SwiftCliCoreTests/InputValidatorTest.swift [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        print("Expected:")
+        print(expected)
+        var IsItTheSame : Bool! = nil
+        if (result == expected){
+            IsItTheSame = true
+        }
+        else{
+            IsItTheSame = false
+        }
+        XCTAssertEqual(IsItTheSame, true)
```
Comment (hnlee) [inline Tests/SwiftCliCoreTests/InputValidatorTest.swift:61]: Hm, I'm a little confused why you couldn't just use `XCTAssertEqual(result, expected)`?

**Review by hnlee — CHANGES_REQUESTED** [review]
Great work overall! I made some suggestions here and there, but I think you're off to a good start!

**Comment by breadoliveoilsalt** [comment]
Great start! Agree with Hana on her comments.  Another thought I had is for a future story, not needed for this PR: to beef up the README to provide instructions to another dev on how to run the code base locally.  I'm thinking selfishly -- how do I download your code and get this up and running? How do I run the tests? What dependencies (or other stuff) do I need to download to do this?  I tried to find some examples among the 8L repos...This one seemed pretty good.  Although yours might not have to be this fulsome...

https://github.com/8thlight/lom_a_backend/blob/main/README.md

File: Sources/SwiftCliCore/ChartConstructor.swift [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+                finished_chart += self.make_stitch_row(row: stitch_array[row])
+            }
+            //if non-first row
+            else {
+                finished_chart += self.make_middle_row(width: row_length)
+                finished_chart += self.make_stitch_row(row: stitch_array[row])
+                //if last row
+
+            }
+            if (row == (number_of_rows - 1)){
```
[the remark this answers is not in the document]
Comment (graceolivia) [reply Sources/SwiftCliCore/ChartConstructor.swift:28]: Makes a lot more sense! I've already been confused thinking about rows in terms of starting at 0 or 1, using that variable name is so clarifying.

File: Sources/SwiftCliCore/InputValidator.swift [hunk 5]
```diff
@@ -0,0 +1,29 @@
+
+import Foundation
+
+// validator class
+public class Validator {
```
[the remark this answers is not in the document]
Comment (graceolivia) [reply Sources/SwiftCliCore/InputValidator.swift:5]: I renamed it!

File: Sources/SwiftCliCore/InputValidator.swift [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+            }
+        }
+        return true
+    }
+    public func ArrayMaker(cleanedpattern: String) -> [[String]] {
+        var stitch_array: [[String]] = []
+        let pattern_rows = cleanedpattern.components(separatedBy: "\n")
+        for row in pattern_rows{
+            var pattern_stitches = row.components(separatedBy: " ")
+            pattern_stitches.removeAll(where: { $0 == "" })
```
[the remark this answers is not in the document]
Comment (graceolivia) [reply Sources/SwiftCliCore/InputValidator.swift:23]: Nice! I implemented this. I did end up having to convert the substrings `splits()` produces into strings, but then I was able to delete several lines of code.

File: Tests/SwiftCliCoreTests/InputValidatorTest.swift (hunk 3 above)
Comment (graceolivia) [reply Tests/SwiftCliCoreTests/InputValidatorTest.swift:61]: I don't know why I didn't do that. I fixed it!

**Review by hnlee — APPROVED** [review]
Great work!

**Pull request merged**